sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/objects/objects.scala

```diff
@@ -1261,8 +1261,24 @@ case class InitializeJavaBean(beanInstance: Expression, setters: Map[String, Exp
   override def children: Seq[Expression] = beanInstance +: setters.values.toSeq
   override def dataType: DataType = beanInstance.dataType
 
-  override def eval(input: InternalRow): Any =
-    throw new UnsupportedOperationException("Only code-generated evaluation is supported.")
+  override def eval(input: InternalRow): Any = {
+    val instance = beanInstance.eval(input).asInstanceOf[Object]
+    if (instance != null) {
+      setters.foreach { case (setterMethod, fieldExpr) =>
```

Comment:
Why are we resolving setters during `eval`? That seems a bit expensive. How about we create the setters before we execute eval? For example (I got a bit carried away):
```scala
  private lazy val resolvedSetters = {
    val ObjectType(beanClass) = beanInstance.dataType
    val lookup = MethodHandles.lookup()
    setters.map {
      case (name, expr) =>
        // Resolve expression type (should be better!)
        val fieldClass = CallMethodViaReflection.typeMapping(expr.dataType).head
        val handle = lookup.findVirtual(
          beanClass,
          name,
          MethodType.methodType(classOf[Unit], fieldClass))
        handle -> expr
    }
  }

  override def eval(input: InternalRow): Any = {
    val bean = beanInstance.eval(input)
    resolvedSetters.foreach {
      case (setter, expr) =>
        setter.invoke(bean, expr.eval(input))
    }
    bean
  }
```

Reply:
`lookup.findVirtual` or `lookup.findSetter`? Strictly I think `InitializeJavaBean` should be used only for setters not for general methods.

Reply:
`lookup.findSetter` means direct field access in the `MethodHandle` world.  So you need to use `lookup.findVirtual`.

Reply:
Another question: Do we need to find method with argument types? Codegen execution also does not do this check. I would think the caller of `InitializeJavaBean` should give correct arguments to the setters.

Reply:
Yes, we do. The compiler will check the generated code right? It is just a matter of where you decide to fail. In a perfect world this would happen on the driver.

Reply:
This condition `MethodType.methodType(classOf[Unit], fieldClass))` is more strict than codegen execution. Codegen execution does not care about return type. So the method such as `def add(x$1: Int): Boolean` of `java.util.LinkedList.add` can work for codegen version. But it fails the method looking here.

Reply:
Hmmm it should be lenient. Can you try `MethodType.methodType(classOf[Object], fieldClass))`?

Reply:
Or should we go back to reflection?

Reply:
See https://github.com/apache/spark/pull/20753#issuecomment-371195614 :)